File: doc/codes/csharp/examples/tutorial/single-device.cs
using System;
using System.Numerics;
using System.Threading;
using System.Threading.Tasks;
using AUTD3;
using AUTD3.Link;
using static AUTD3.Units;

// HIDE
namespace DocSamples.TutorialSingleDevice;

internal static class Sample
{
    internal static async Task Run()
    {
        // HIDE_END
// Define a geometry consisting of a single AUTD3 device.
var geometry = new Geometry(new[] { new Autd3(Vector3.Zero) });

// Open the client over an echocat link.
await using var client = await Client.OpenAsync(
    geometry,
    new EchocatLinkOption(),
    new ClientConfig()
);

// Generate a focus 150 mm above the array center.
var target = geometry.Center + new Vector3(0.0f, 0.0f, 150.0f);
var wavelength = Pattern.Wavelength(340.0f * m / s);
var patterns = geometry.PatternBuffer();
Pattern.Focus(
    geometry,
    target,
    wavelength,
    new FocusOption(),
    patterns
);

// Apply a 200 Hz sine-wave AM.
var modulation = Modulation.ModulationBuffer();
Modulation.Sine(
    200 * Hz,
    new SineOption(samplingConfig: SamplingConfig.Freq4k),
    modulation
);

var builder = client.DatagramBuilder();
builder.Push(new SetSilencer());
builder.Push(new Pattern(patterns));
builder.Push(new Modulation(SamplingConfig.Freq4k, modulation));
var frames = builder.Build();
foreach (var frame in frames)
{
    await client.SendCheckedAsync(frame);
}

using var cts = new CancellationTokenSource();
Console.CancelKeyPress += (_, e) =>
{
    e.Cancel = true;
    cts.Cancel();
};

try
{
    await Task.Delay(Timeout.InfiniteTimeSpan, cts.Token);
}
catch (OperationCanceledException)
{
}
        // HIDE
    }
}
// HIDE_END

File: doc/codes/python/examples/tutorial/single-device.py
import asyncio

import numpy as np

import autd3_link_echocat as echocat
import autd3_modulation as modulation
import autd3_pattern as pattern
from autd3 import Client, ClientConfig
from autd3.commands import Modulation, Pattern, SetSilencer
from autd3.geometry import Autd3, Geometry
from autd3.units import Hz, m, s
from autd3.value import SamplingConfig

# xtask:long-running  # [hide]


async def main() -> None:
    # Define a geometry consisting of a single AUTD3 device.
    geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    # Open the client over an echocat link.
    async with await Client.open(
        geometry,
        echocat.EchocatLinkOption(),
        ClientConfig(),
    ) as client:
        # Generate a focus 150 mm above the array center.
        target = geometry.center() + np.array([0.0, 0.0, 150.0])
        wavelength = pattern.wavelength(340 * m / s)
        patterns = geometry.pattern_buffer()
        pattern.focus(
            geometry,
            target,
            wavelength,
            pattern.FocusOption(),
            patterns,
        )

        # Apply a 200 Hz sine-wave AM.
        mod_buf = modulation.modulation_buffer()
        modulation.sine(
            200 * Hz,
            modulation.SineOption(sampling_config=SamplingConfig.FREQ_4K),
            mod_buf,
        )

        builder = client.datagram_builder()
        builder.push(SetSilencer())
        builder.push(Pattern(patterns))
        builder.push(Modulation(SamplingConfig.FREQ_4K, mod_buf))
        for frame in builder.build():
            await client.send_checked(frame)

        await asyncio.Event().wait()


if __name__ == "__main__":
    asyncio.run(main())
